keps/sig-storage/2317-fsgroup-on-mount/README.md

```diff
@@ -0,0 +1,400 @@
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+## Production Readiness Review Questionnaire
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+
+-->
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: MountWithFSGroup
+    - Components depending on the feature gate:
+      - Kubelet
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+  Enabling this feature on its own does not changes any existing behaviour but if
+  selected CSI driver advertise `VOLUME_MOUNT_GROUP` node capability, it will result in
+  volume being mounted with pod's `fsGroup`.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Yes - feature gate can be disabled once enabled and kubelet will fallback to its current behavior, which will
+  result in any CSI driver that requires using `fsGroup` as a mount option to not use the mount option. In which case
+  mounted volumes will be unwritable by any pods other than those running as root.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  It will cause kubelet to use `volume_mount_group` field of CSI whenver applicable as discussed in above design.
+
+* **Are there any tests for feature enablement/disablement?**
+  Not yet.
+
+### Rollout, Upgrade and Rollback Planning
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can a rollout fail? Can it impact already running workloads?**
```

Comment:
This question seems important for this feature. Does the scheduler have any way to know whether a particular node will honor this part of the podspec?

Reply:
I do not think scheduler has to know and it can be pretty expensive(code and complexity perspective) to bubble this up to scheduler.

Reply:
> I do not think scheduler has to know and it can be pretty expensive(code and complexity perspective) to bubble this up to scheduler.

So that means that mid-rollout (either turning it on or turning it off) can result in different pods of the same replicaset/deployment failing or succeeding in seemingly random ways?  I think that's worth noting in an answer to this question.

Reply:
hmm - if the replicaset was say working in 1.20 - then user must be running the pods as root or mounting the volume as world-writable (`777`). So this feature won't affect that. Even if `fsGroup` is applied correct to the volume, it won't affect anything that is running as root (because those pods can write to any volume anyways) or volumes with 777 permissions (because those volumes can be written by anyone anyways).

So in a nutshell - it does not have any impact of already correctly running workloads. Now if you somehow had a pod in 1.20 - which was not running as root and volumes weren't mounted with 777 permissions then those would suddenly start working. But I do not think that is the use case this section documents.